### ISSUE TITLE
How to generate train.tfrecord?

### ISSUE THREAD

**Issue by cwl1999** [deepmind/deepmind-research#204]
Dear author:
           Hello! I am a graduate student from Wuhan Textile University in China. I am studying your project about learning recently_ to_ simulate. I'm very interested in this. Can you provide the generated data train.tfrecord Source dataset file for? My email address is: cwl_0513@163.com.
           Thank you very much for your reading and look forward to your reply.

**Comment by alvarosg** [comment]
Thank you for your message? 

> Can you provide the generated data train.tfrecord Source dataset file for?

Could you clarify the question/which datasets you are referring to? All datasets should be downloadable by name using the download script, as per the instructions in the README:

``` For example
mkdir -p /tmp/datasets
bash ./learning_to_simulate/download_dataset.sh WaterRamps /tmp/datasets
```

Hope this helps.

**Comment by cwl1999** [comment]
Thank you very much for your answer! Because I want to try to make a cloth dataset for simulation experiments, but it is difficult for me to obtain the data content of tf.records file. When I forcibly open it, I can only see the garbled code. I'm confused now. Please give me some advice.

**Comment by alvarosg** [comment]
Thanks for your reply. If you want human readable access, I would recommend using python to read the dataset and iterate through the examples. You may use TF2 to do this:

```
    import tensorflow as tf
    from learning_to_simulate import reading_utils

    ds = tf.data.TFRecordDataset([os.path.join(data_path, f'{split}.tfrecord')])
    ds = ds.map(functools.partial(
        reading_utils.parse_serialized_simulation_example, metadata=metadata))

    for element in ds.as_numpy_iterator():
      print(element)
```

**Comment by aakash30jan** [comment]
> ```
>     for element in ds.as_numpy_iterator():
> ```

 ```
     for element in ds:
```

**Comment by Social-Mean** [comment]
I have a simular question. I want to creat a test dataset with ramps set by myself. What should I do? How can I creater such a `test.tfrecord` file?

**Comment by oasis-asu** [comment]
> Thanks for your reply. If you want human readable access, I would recommend using python to read the dataset and iterate through the examples. You may use TF2 to do this:
> 
> ```
>     import functools
>     import tensorflow as tf
>     from learning_to_simulate import reading_utils
> 
>     ds = tf.data.TFRecordDataset([os.path.join(data_path, f'{split}.tfrecord')])
>     ds = ds.map(functools.partial(
>         reading_utils.parse_serialized_simulation_example, metadata=metadata))
> 
>     for element in ds.as_numpy_iterator():
>       print(element)
> ```

What is metadata? Do I need to change reading_utils according to different dataset?

**Comment by BoyuanTang331** [comment]
> > Thanks for your reply. If you want human readable access, I would recommend using python to read the dataset and iterate through the examples. You may use TF2 to do this:
> > ```
> >     import functools
> >     import tensorflow as tf
> >     from learning_to_simulate import reading_utils
> > 
> >     ds = tf.data.TFRecordDataset([os.path.join(data_path, f'{split}.tfrecord')])
> >     ds = ds.map(functools.partial(
> >         reading_utils.parse_serialized_simulation_example, metadata=metadata))
> > 
> >     for element in ds.as_numpy_iterator():
> >       print(element)
> > ```
> 
> What is metadata? Do I need to change reading_utils according to different dataset?

like a global parameter

**Comment by yours612** [comment]
Hello, I want to know how vel_mean, vel_std, acc_mean, and acc_std are calculated in metadata? I used one of the deepmind datasets and used the sequence of positions in one of the segments to calculate mean and std, but the calculations were much different from those in metadata

**Comment by alvarosg** [comment]
Thanks for your message!

So note what we call velocity statistics, are more like the statistics of the difference in positions. And similar for the accelerations. Assuming a timestep of 1 for simplicity. (See "Input and output representations" paragraph in the paper)

This means what we call the velocity statistics are really the average across particles, steps, and trajectories of (x_tp1 - x_t) (omitting the timestep!). And similar for the acceleration, but with x_tp1 - 2x_t + x_tm1

Hope this helps!

**Comment by yq60523** [comment]
> Thanks for your message!
> 
> So note what we call velocity statistics, are more like the statistics of the difference in positions. And similar for the accelerations. Assuming a timestep of 1 for simplicity. (See "Input and output representations" paragraph in the paper)
> 
> This means what we call the velocity statistics are really the average across particles, steps, and trajectories of (x_tp1 - x_t) (omitting the timestep!). And similar for the acceleration, but with x_tp1 - 2x_t + x_tm1
> 
> Hope this helps!

Hi Alvaro,

I have some problems with it. The 'step_context' is mentioned in the dataset creation, but it isn't considered in the training of the 'learning to simulate' model. Can you explain this reason? Is this variable important?

**Comment by alvarosg** [comment]
Hi, the global `step_context` is used [here](https://github.com/google-deepmind/deepmind-research/blob/f5de0ede8430809180254ee957abf36ed62579ef/learning_to_simulate/train.py#L121) when preparing the inputs to the model.

**Comment by yq60523** [comment]
> Hi, the global `step_context` is used [here](https://github.com/google-deepmind/deepmind-research/blob/f5de0ede8430809180254ee957abf36ed62579ef/learning_to_simulate/train.py#L121) when preparing the inputs to the model.

Hi Alvaro,
Thanks for your reply. I also found this usage. However, I can't understand what the 'step_context' means. Can I use this variable to store the timestep and global operational parameters? I encountered a problem in the usage of this code, which is that the prediction accuracy doesn't satisfy our requirements. Therefore, I want to use the timestep as a variable in the prediction to improve its accuracy and add the global operational parameters into this model to extend its usage. However, I don't know if my idea will be successful. Have you been able to try this before?

**Comment by alvarosg** [comment]
From the paper this corresponds to global context, e.g. "The global properties of the system, g, include external forces
and global material properties, when applicable."

Then from the supplementary:
"In FLUIDSHAKE, particle positions were provided in the coordinate frame of the container, and the container position, velocity and acceleration were provided as 6 global features. In CONTINUOUS a single global scalar was used to indicate the friction angle of the material."

The rest of the datatsets don't use a global context if I remember correctly.

**Comment by yq60523** [comment]
> From the paper this corresponds to global context, e.g. "The global properties of the system, g, include external forces and global material properties, when applicable."
> 
> Then from the supplementary: "In FLUIDSHAKE, particle positions were provided in the coordinate frame of the container, and the container position, velocity and acceleration were provided as 6 global features. In CONTINUOUS a single global scalar was used to indicate the friction angle of the material."
> 
> The rest of the datatsets don't use a global context if I remember correctly.

Hi Alvaro,

Thanks for your reply. The CONTINUOUS and FLUIDSHAKE datasets contain the 'step_context'. However, I want to know how to save the 'step_context' into the tfrecord file. I have tried many methods, which did not work. The code of the generate_tfrecord_dataset.py doesn't work. I've checked the 'step_context' of the two datasets and found that the first variable of the 'step_context' list is Nan. Does the first variable affect the performance of this model?

**Comment by alvarosg** [comment]
> The code of the generate_tfrecord_dataset.py doesn't work.

Unfortunately generating new datasets was out of the scope of this repo, but hopefully you can find some help from other issues of others trying doing that.

> I've checked the 'step_context' of the two datasets and found that the first variable of the 'step_context' list is Nan. Does the first variable affect the performance of this model?

I believe the model only needs to condition in element at position [-2] like [here](https://github.com/google-deepmind/deepmind-research/blob/f5de0ede8430809180254ee957abf36ed62579ef/learning_to_simulate/train.py#L124), which if I remember correctly corresponds to the last input step (e.g. [-1] would the the context for the target step which you don't usually condition on, and before [-2] it would problably be prior input steps, which in this case are also not necessary, and probalby why it is sometimes left as nan). Is this consistent with what you are seeing?

**Comment by yq60523** [comment]
> > The code of the generate_tfrecord_dataset.py doesn't work.
> 
> Unfortunately generating new datasets was out of the scope of this repo, but hopefully you can find some help from other issues of others trying doing that.
> 
> > I've checked the 'step_context' of the two datasets and found that the first variable of the 'step_context' list is Nan. Does the first variable affect the performance of this model?
> 
> I believe the model only needs to condition in element at position [-2] like [here](https://github.com/google-deepmind/deepmind-research/blob/f5de0ede8430809180254ee957abf36ed62579ef/learning_to_simulate/train.py#L124), which if I remember correctly corresponds to the last input step (e.g. [-1] would the the context for the target step which you don't usually condition on, and before [-2] it would problably be prior input steps, which in this case are also not necessary, and probalby why it is sometimes left as nan). Is this consistent with what you are seeing?

Hi Alvaro,

Thanks for your answer. I'm trying this variable now, but it seems not to perform well, even if in the training dataset. Based on your paper and code, your model outputs the acceleration of the current timestep, and the position is calculated with the acceleration. Why don't you output the position directly, since a little bit of error in the acceleration will lead to a big difference in the position? In my case, the predicted acceleration is usually bigger than the normal, leading to an abnormal position.

**Comment by yq60523** [comment]
> > The code of the generate_tfrecord_dataset.py doesn't work.
> 
> Unfortunately generating new datasets was out of the scope of this repo, but hopefully you can find some help from other issues of others trying doing that.
> 
> > I've checked the 'step_context' of the two datasets and found that the first variable of the 'step_context' list is Nan. Does the first variable affect the performance of this model?
> 
> I believe the model only needs to condition in element at position [-2] like [here](https://github.com/google-deepmind/deepmind-research/blob/f5de0ede8430809180254ee957abf36ed62579ef/learning_to_simulate/train.py#L124), which if I remember correctly corresponds to the last input step (e.g. [-1] would the the context for the target step which you don't usually condition on, and before [-2] it would problably be prior input steps, which in this case are also not necessary, and probalby why it is sometimes left as nan). Is this consistent with what you are seeing?

Hi Alvaro.
Sorry for disturbing you again. I still have a question about the dataset generation. In the metadata.json, the Δt is also recorded. However, it isn't used in the training code. In contrast, in the position computation of the training code, Δt equals 1.  I want to know why you consider this, and what Δt you use to calculate the mean and std of the velocity and acceleration.

**Comment by alvarosg** [comment]
> In contrast, in the position computation of the training code, Δt equals 1. 

This is because to keep things simple, when we trained the model we technically train it to predict a different between positions, rather than a velocity. This is sort of equivalent to saying that we train the model to predict velocity assuming the timestep was 1, instead of whatever it is. So long as your timestep is the same across all the datasets, that each model trains, the units of time should not matter, so you don't lose any generality by doing it this way. Of course if different datasets that are being used to train the same model had different timestamps, then the timestep should be rescaled in a consistent way across them, rather than simply using it as 1 for all of them.

**Comment by yq60523** [comment]
> > In contrast, in the position computation of the training code, Δt equals 1.
> 
> This is because to keep things simple, when we trained the model we technically train it to predict a different between positions, rather than a velocity. This is sort of equivalent to saying that we train the model to predict velocity assuming the timestep was 1, instead of whatever it is. So long as your timestep is the same across all the datasets, that each model trains, the units of time should not matter, so you don't lose any generality by doing it this way. Of course if different datasets that are being used to train the same model had different timestamps, then the timestep should be rescaled in a consistent way across them, rather than simply using it as 1 for all of them.

So you mean that the Δt used in the velocity calculations in the metadata.json and in the training code should be consistent. I just want to make sure about this. I tried the physical running time previously without modifying the training code. However, it doesn't perform well.

**Comment by alvarosg** [comment]
What I mean is that is that if you train the model to predict `target = (next_x - prev_x) / delta_t_dataset`, then when you run it you should compute `next_predicted_x = model_output * delta_t_dataset + prev_x`.

But you can also choose to use different time unites, and say: new_time = time / delta_t_dataset, so then you can train the model to predict `target = (next_x - prev_x) / 1`, then when you run your model, you should compute `next_predicted_x = model_output * 1 + prev_x`. (e.g. in this case, we re-normalize time to have an effective timestep of 1 without losing generlizaty).

But if you are training the same model (same set of weights) in two datasets, one with delta_t_dataset_1 and another with a different delta_t_dataset_2, then you cannot set it to 1 for both, then you either normalize time by the delta_t of the first dataset, or the one of the second dataset.

**Comment by yq60523** [comment]
> What I mean is that is that if you train the model to predict `target = (next_x - prev_x) / delta_t_dataset`, then when you run it you should compute `next_predicted_x = model_output * delta_t_dataset + prev_x`.
> 
> But you can also choose to use different time unites, and say: new_time = time / delta_t_dataset, so then you can train the model to predict `target = (next_x - prev_x) / 1`, then when you run your model, you should compute `next_predicted_x = model_output * 1 + prev_x`. (e.g. in this case, we re-normalize time to have an effective timestep of 1 without losing generlizaty).
> 
> But if you are training the same model (same set of weights) in two datasets, one with delta_t_dataset_1 and another with a different delta_t_dataset_2, then you cannot set it to 1 for both, then you either normalize time by the delta_t of the first dataset, or the one of the second dataset.

I got it. Thanks very much for your help.

**Comment by yq60523** [comment]
> What I mean is that is that if you train the model to predict `target = (next_x - prev_x) / delta_t_dataset`, then when you run it you should compute `next_predicted_x = model_output * delta_t_dataset + prev_x`.
> 
> But you can also choose to use different time unites, and say: new_time = time / delta_t_dataset, so then you can train the model to predict `target = (next_x - prev_x) / 1`, then when you run your model, you should compute `next_predicted_x = model_output * 1 + prev_x`. (e.g. in this case, we re-normalize time to have an effective timestep of 1 without losing generlizaty).
> 
> But if you are training the same model (same set of weights) in two datasets, one with delta_t_dataset_1 and another with a different delta_t_dataset_2, then you cannot set it to 1 for both, then you either normalize time by the delta_t of the first dataset, or the one of the second dataset.

Hi Alvaro,

A problem exists that can't be solved when I start to use this model, which is the error accumulation. When I try to train this model your 3D dataset and my own dataset, the error accumulation usually exists. In the first few steps of testing, this model works very well. However, the error accumulates with the increasing steps. I know your paper discussed the effects of the noise on this problem. However, this approach shows little improvement in my testing. Therefore, I'm asking for your advice.

**Comment by nsrawat0333** [comment]
Hi @alvarosg and team,

I've created a complete solution for this TFRecord generation issue that addresses all the questions raised in this conversation thread.

**Solution includes:**
• `generate_tfrecord_dataset.py` - Complete TFRecord generation pipeline with automatic statistics calculation
• `tfrecord_reader_example.py` - Human-readable TFRecord inspection tool (solves "garbled code" issue)
• `TFRECORD_GENERATION_GUIDE.md` - Comprehensive documentation with examples

**Key implementations:**
- Statistics calculation following @alvarosg's explanation: `velocities = positions[1:] - positions[:-1]` (Δt=1)
- Cloth simulation dataset creation (addresses @cwl1999's original question)
- TFRecord format explanation (binary format, not human-readable when opened directly)
- step_context support and error handling

**Usage:**
```bash
# Create sample cloth dataset
python generate_tfrecord_dataset.py --create_sample --output_dir=cloth_dataset

# Read TFRecord contents (no garbled code!)
python tfrecord_reader_example.py --tfrecord_path=train.tfrecord

# Train model
python -m learning_to_simulate.train --data_path=cloth_dataset --model_path=models/
